src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,442 @@
+                                let _x = sender.send(Ok(chunk));
+                                let _x = sender.send(Err(e));
+                                break 'chunk_loop;
+                            }
+                        }
+                    }
+                    // If there is nothing in the chunk because the innner iterator is
+                    // exhausted, or we get a send error (implying the receiver has been
+                    // dropped), then exit the thread's main loop.
+                    if chunk.is_empty() || sender.send(Ok(chunk)).is_err() {
+                        break;
+                    }
+                }
+            })
+            .expect("failed to spawn chunked read ahead thread");
+
+        // Store the necessary objects on ``Self``
+        Self { receiver, join_handle: Some(join_handle), current_chunk: Vec::new().into_iter() }
+    }
+}
+
+impl<T> Iterator for ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        // Check if our current chunk has anything left in it
+        // If so, just return that result
+        // If not, see documentation on else block
+        let next_option = self.current_chunk.next();
+        if next_option.is_some() {
+            next_option
+        } else {
+            // Current chunk didn't have anything left in it, so
+            // Try to grab a new chunk, and panic if there are no chunks left (note that
+            // ``recv`` is blocking, so this will only return an error if the sender has been
+            // dropped and there are no more elements in the channel.)
+            // Clippy incorrectly marks this as something that can be done with a question mark
+            // so ignore that lint
+
+            #[allow(clippy::question_mark)]
+            let res_r = if let Ok(result) = self.receiver.recv() {
+                result
+            } else {
+                // join handle is not ``Copy`` or ``Clone`` and we need ownership of it to be able
+                // to join on it, hence the optional field and taking it off the iterator struct.
+                if let Some(join_handle) = self.join_handle.take() {
+                    if let Err(e) = join_handle.join() {
+                        resume_unwind(e)
+                    }
+                }
+                return None;
+            };
+            // If the new chunk is present and Ok, convert it to an iterator, store it on ``self``,
+            // and return its next value ( shutting down our reciever if the next value is None).
+            // if the new chunk is an Err, raise it to the main thread.
+            match res_r {
+                Ok(next_chunk) => {
+                    self.current_chunk = next_chunk.into_iter();
+                    self.current_chunk.next()
+                }
+                Err(e) => {
+                    resume_unwind(e);
+                }
+            }
```

Comment:
I still find this whole block really hard to follow. 

E.g. why do `let res_r = if let ...` when the `else` branch returns from the function?

I know if/let is common in rust, but when we have lots of layers of results/options going on I think case/match might be clearer.  

At a minimum I would suggest:

```suggestion
            if let Ok(chunk_or_panic) = self.receiver.recv() {
                // If the new chunk is present and Ok, convert it to an iterator, store it on ``self``,
                // and return its next value ( shutting down our reciever if the next value is None).
                // if the new chunk is an Err, raise it to the main thread.
                match chunk_or_panic {
                    Ok(next_chunk) => {
                        self.current_chunk = next_chunk.into_iter();
                        self.current_chunk.next()
                    }
                    Err(e) => {
                        resume_unwind(e);
                    }
                }
            } else {
                // join handle is not ``Copy`` or ``Clone`` and we need ownership of it to be able
                // to join on it, hence the optional field and taking it off the iterator struct.
                if let Some(join_handle) = self.join_handle.take() {
                    if let Err(e) = join_handle.join() {
                        resume_unwind(e)
                    }
                }
                return None;
            };
```